modules/cluster_estimation/cluster_estimation_by_label.py

```diff
@@ -0,0 +1,177 @@
+        run_override: bool,
+    ) -> "tuple[True, dict[int, list[object_in_world.ObjectInWorld]]] | tuple[False, None]":
+        """
+        Take in list of detections and return list of estimated object locations
+        if number of detections is sufficient, or if manually forced to run.
+
+        PARAMETERS
+        ----------
+        input_detections: list[DetectionInWorld]
+            List containing DetectionInWorld objects which holds real-world positioning data to run
+            clustering on.
+
+        run_override: bool
+            Forces ClusterEstimation to predict if data is available, regardless of any other
+            requirements.
+
+        RETURNS
+        -------
+        model_ran: bool
+            True if ClusterEstimation object successfully ran its estimation model, False otherwise.
+
+        labels_to_object_clusters: dict[int, list[object_in_world.ObjectInWorld] or None.
+            Dictionary where the key is a label and the value is a list of all cluster detections with that label
+        """
+        label_to_detections: dict[int, list[detection_in_world.DetectionInWorld]] = {}
+        # Sorting detections by label
+        for detection in input_detections:
+            if not detection.label in label_to_detections:
+                label_to_detections[detection.label] = []
+            label_to_detections[detection.label].append(detection)
+
+        labels_to_object_clusters: dict[int, list[object_in_world.ObjectInWorld]] = {}
+        for label, detections in label_to_detections.items():
+            # create cluster estimation for label if it doesn't exist
+            if not label in self.__label_to_cluster_estimation_model:
+                result, cluster_model = cluster_estimation.ClusterEstimation.create(
+                    self.__min_activation_threshold,
+                    self.__min_new_points_to_run,
+                    self.__max_num_components,
+                    self.__random_state,
+                    self.__local_logger,
+                )
+                if not result:
+                    self.__local_logger.error(
+                        f"Failed to create cluster estimation for label {label}"
+                    )
+                    return False, None
+                self.__label_to_cluster_estimation_model[label] = cluster_model
+            # runs cluster estimation for specific label
+            result, clusters = self.__label_to_cluster_estimation_model[label].run(
+                detections,
+                run_override,
+            )
+            if not result:
+                self.__local_logger.error(
+                    f"Failed to run cluster estimation model for label {label}"
+                )
+                return False, None
+            if not label in labels_to_object_clusters:
+                labels_to_object_clusters[label] = []
+            labels_to_object_clusters[label] += clusters
```

Comment:
This doesn't seem necessary since the labels are guaranteed unique in the for loop. Just do an assignment. Add a comment with this explanation.